HiP-ThumbnailService/Controllers/ThumbnailsController.cs

```diff
@@ -0,0 +1,160 @@
+
+            var encodedId = Convert.ToBase64String(Encoding.UTF8.GetBytes(args.Url));
+            var folderPath = Path.Combine(_dirConfig.Path, encodedId);
+
+            if (string.IsNullOrEmpty(args.Size) || !_sizeConfig.SupportedSizes.ContainsKey(args.Size))
+                return BadRequest(new { Message = "Invalid size" });
+
+            var semaphore = LockDictionary.GetOrAdd(encodedId, new SemaphoreSlim(1));
+            await semaphore.WaitAsync();
+            try
+            {
+                var extension = RequestImageFormatUtils.GetExtension(args.Format);
+                var thumbnailPath = Path.Combine(folderPath, GetFileName(args.Size, args.Mode)) + "." +
+                                    extension;
+                if (!System.IO.File.Exists(thumbnailPath))
+                {
+                    var client = new HttpClient();
+                    client.DefaultRequestHeaders.Add("Authorization", Request.Headers["Authorization"].ToString());
+                    var hostUrl = _endpointConfig.HostUrl;
+                    var stream = await client.GetStreamAsync(hostUrl + args.Url);
+
+                    if (stream == null)
+                    {
+                        return BadRequest(new { Message = "Image access failed" });
+                    }
+
+                    //Create Directory if it doesn't exist
+                    if (!Directory.Exists(folderPath))
+                        Directory.CreateDirectory(folderPath);
+
+                    var value = _sizeConfig.SupportedSizes[args.Size];
+                    GenerateThumbnail(args.Mode, stream, thumbnailPath, value);
+                }
+
+                return File(new FileStream(thumbnailPath, FileMode.Open), $"image/{extension}",
+                        Path.GetFileName(thumbnailPath));
+
+            }
+            finally
+            {
+                semaphore.Release();
+            }
+
+        }
+
+        /// <summary>
+        /// Generates a thumbnail, if it doesn't exist yet
+        /// </summary>
+        /// <param name="mode">Crop mode</param>
+        /// <param name="fileStream">File stream</param>
+        /// <param name="thumbnailPath">Path to thumbnail</param>
+        /// <param name="value">Size value</param>
+        private static void GenerateThumbnail(CropMode mode, Stream fileStream, string thumbnailPath, int value)
+        {
+            switch (mode)
+            {
+                case CropMode.FillSquare:
+
+                    using (var image = Image.Load(fileStream))
+                    {
+                        image.Mutate(c =>
+                            c.Resize(new ResizeOptions { Mode = ResizeMode.Crop, Size = new Size(value) }));
+                        image.Save(thumbnailPath);
+                    }
+                    break;
+                case CropMode.Uniform:
+                    using (var image = Image.Load(fileStream))
```

Comment:
Duplicate code, the only difference is `ResizeMode.Crop` vs `ResizeMode.Max`. You can write this more succinctly.